Updated ZMQ implementation to use NULL if scope is not set. Only when…

abroekhuis opened the pull request:
… needed "default" is used.

Added exclude filter to example for scope, to make sure that the right publisher is found. If a publisher without scope, and a publisher with a scope, use the same topic, this could result in getting the wrong publisher service.

For now, only updated the ZMQ admin.

Oipo commented on bundles/pubsub/pubsub_api/include/pubsub/subscriber.h:
```diff
@@ -39,7 +39,7 @@
 #define PUBSUB_SUBSCRIBER_CONFIG               "pubsub.config"
 
-#define PUBSUB_SUBSCRIBER_SCOPE_DEFAULT        "default"
+//#define PUBSUB_SUBSCRIBER_SCOPE_DEFAULT        "default"
```
Why comment out the subscriber but not the publisher?

abroekhuis replied:
Leftovers, fixed.

abroekhuis commented on bundles/pubsub/pubsub_admin_zmq/src/pubsub_zmq_admin.c:
```diff
@@ -667,13 +667,17 @@ static celix_status_t pubsub_zmqAdmin_connectEndpointToReceiver(pubsub_zmq_admin
+                pubsub_zmqTopicReceiver_connectTo(receiver, url);
+            } else if (scope != NULL && eScope != NULL && strncmp(scope, eScope, 1024*1024) == 0) {
+                pubsub_zmqTopicReceiver_connectTo(receiver, url);
+            }
```
Not needed, since a discovered endpoint might not have a user, and as such, no receiver.

abroekhuis commented on bundles/pubsub/pubsub_discovery/src/pubsub_discovery_impl.c:
```diff
@@ -414,7 +414,7 @@ celix_status_t pubsub_discovery_announceEndpoint(void *handle, const celix_prope
         entry->isSet = false;
         entry->properties = celix_properties_copy(endpoint);
-        asprintf(&entry->key, "/pubsub/%s/%s/%s/%s", config, scope, topic, uuid);
+        asprintf(&entry->key, "/pubsub/%s/%s/%s/%s", config, scope == NULL ? "default" : scope, topic, uuid);
```
Agreed, fixed where needed. Moved the constant to pubsub_constants.

abroekhuis commented on bundles/pubsub/pubsub_spi/src/pubsub_endpoint.c:
```diff
@@ -194,7 +196,11 @@ bool pubsubEndpoint_equals(const celix_properties_t *psEp1, const celix_properti
+    if (scope != NULL) {
+        asprintf(&result, "%s:%s", scope, topic);
+    } else {
+        asprintf(&result, "default:%s", topic);
```
Fixed

Oipo replied:
Think you missed this one ;)

abroekhuis commented on bundles/pubsub/pubsub_spi/src/pubsub_utils.c:
```diff
@@ -92,7 +78,7 @@ celix_status_t pubsub_getPubSubInfoFromFilter(const char* filterstr, char **topi
     array_list_pt bundles = NULL;
     bundleContext_getBundles(ctx, &bundles);
-    int nrOfBundles = arrayList_size(bundles);
+    uint32_t nrOfBundles = arrayList_size(bundles);
```
I agree, but should not be part of this PR. Going to leave it as it is now.

abroekhuis commented:
Updated the other admins to allow a NULL scope as well. Left out the nano pubsubadmin, as it is experimental, and needs more work.

@ErjanAltena can the nano pubsubadmin be moved to a branch, and removed from develop? I don't think it is ready yet for develop and later a release.

Oipo commented on bundles/pubsub/pubsub_admin_zmq/src/pubsub_zmq_admin.c:
```diff
@@ -685,9 +658,12 @@ celix_status_t pubsub_zmqAdmin_addDiscoveredEndpoint(void *handle, const celix_p
-            pubsub_zmq_topic_receiver_t *receiver = hashMapIterator_nextValue(&iter);
+        const char *scope = celix_properties_get(endpoint, PUBSUB_ENDPOINT_TOPIC_SCOPE, NULL);
+        const char *topic = celix_properties_get(endpoint, PUBSUB_ENDPOINT_TOPIC_NAME, NULL);
+        char *key = pubsubEndpoint_createScopeTopicKey(scope, topic);
```
Missing free for key

Oipo commented:
Travis doesn't build because of a syntax error on pubsub_tcp_admin.c:581.

abroekhuis commented:
> Travis doesn't build because of a syntax error on pubsub_tcp_admin.c:581.
> @pnoltes I think the github build.yml misses BUILD_PUBSUB_PSA_TCP=ON

I've added a remark to the PR for the updated workflow. Seems like the right place to fix this.
